File: oxidize-python/oxidize_python/core/model/dflash_heuristic.py
"""DFlash heuristic draft mirroring oxidize-golang/core/model/dflash_heuristic.go."""

from __future__ import annotations

import math
from dataclasses import dataclass

from oxidize_python.core.model.dflash import DFlashConfig
from oxidize_python.core.model.model import Logits, Model, Session, Token
from oxidize_python.core.model.sampling import SamplingConfig, _softmax, sample


@dataclass
class DFlashDecodeOpts:
    draft_steps: int = 4
    max_acceptance_rate: float = 0.95
    min_draft_confidence: float = 0.1
    temperature: float = 1.0


def default_dflash_decode_opts() -> DFlashDecodeOpts:
    return DFlashDecodeOpts()


@dataclass
class DFlashStats:
    drafts_generated: int = 0
    drafts_accepted: int = 0
    speedup_estimate: float = 0.0
    avg_draft_latency: float = 0.0

    def acceptance_rate(self) -> float:
        if self.drafts_generated == 0:
            return 0.0
        return self.drafts_accepted / self.drafts_generated


class HeuristicDFlashDraft:
    def __init__(
        self,
        target: Model,
        config: DFlashConfig,
        decode: DFlashDecodeOpts | None = None,
    ) -> None:
        self.target = target
        self.config = config
        self.decode = decode or default_dflash_decode_opts()
        if self.decode.draft_steps <= 0:
            self.decode.draft_steps = config.block_size or 4
        self.stats = DFlashStats()
        self.cache: list[Token] = []

    def forward(self, tokens: list[Token], session: Session) -> Logits:
        return self.target.forward(tokens, session)

    def vocab_size(self) -> int:
        return self.target.vocab_size()

    def context_size(self) -> int:
        return self.target.context_size()

    def layer_count(self) -> int:
        return self.target.layer_count()

    def reset(self) -> None:
        self.cache.clear()

    def generate_draft(self, prompt: list[Token], steps: int = 0) -> list[Token]:
        if steps <= 0:
            steps = self.decode.draft_steps
        out = list(prompt)
        sess = Session()
        for _ in range(steps):
            logits = self.target.forward(out, sess)
            tok = sample(
                logits,
                SamplingConfig(temperature=self.decode.temperature, top_p=1.0),
                None,
            )
            out.append(tok)
            self.cache.append(tok)
        self.stats.drafts_generated += steps
        return out[len(prompt) :]

    def verify_drafts(self, drafts: list[Token]) -> list[Token]:
        if not drafts:
            return []
        sess = Session()
        logits = self.target.forward(drafts, sess)
        probs = _softmax(logits)
        accepted: list[Token] = []
        for t in drafts:
            prob = probs[int(t)] if int(t) < len(probs) else 0.0
            if prob >= self.decode.min_draft_confidence:
                accepted.append(t)
        self.stats.drafts_accepted += len(accepted)
        return accepted


def draft_confidence(target: Model, drafts: list[Token]) -> list[float]:
    if not drafts:
        return []
    sess = Session()
    logits = target.forward(drafts, sess)
    probs = _softmax(logits)
    return [probs[int(t)] if int(t) < len(probs) else 0.0 for t in drafts]


class DFlashEngine:
    def __init__(
        self,
        target: Model,
        draft: HeuristicDFlashDraft,
        config: DFlashConfig,
    ) -> None:
        self.target = target
        self.draft = draft
        self.config = config
        self.decode = draft.decode
        self.stats = DFlashStats()

    def step(self, prompt: list[Token]) -> list[Token]:
        drafts = self.draft.generate_draft(prompt, self.decode.draft_steps)
        accepted = self.draft.verify_drafts(drafts)
        ratio = self.stats.acceptance_rate()
        self.stats.speedup_estimate = 1 + math.ceil(self.decode.draft_steps * ratio)
        return accepted

    def max_acceptance_rate(self) -> float:
        return self.decode.max_acceptance_rate

    def is_stable(self) -> bool:
        return self.stats.acceptance_rate() <= self.decode.max_acceptance_rate

File: oxidize-golang/core/model/dflash_heuristic.go
package model

import (
	"math"
)

// DFlashDecodeOpts tunes the heuristic draft wrapper (target-delegating stub).
type DFlashDecodeOpts struct {
	DraftSteps         int
	MaxAcceptanceRate  float32
	MinDraftConfidence float32
	Temperature        float32
}

// DefaultDFlashDecodeOpts returns balanced speculative-decode defaults.
func DefaultDFlashDecodeOpts() DFlashDecodeOpts {
	return DFlashDecodeOpts{
		DraftSteps:         4,
		MaxAcceptanceRate:  0.95,
		MinDraftConfidence: 0.1,
		Temperature:        1.0,
	}
}

// DFlashStats tracks heuristic draft verification metrics.
type DFlashStats struct {
	DraftsGenerated int
	DraftsAccepted  int
	SpeedupEstimate float32
	AvgDraftLatency float32
}

// AcceptanceRate returns the ratio of accepted drafts to total.
func (s DFlashStats) AcceptanceRate() float32 {
	if s.DraftsGenerated == 0 {
		return 0
	}
	return float32(s.DraftsAccepted) / float32(s.DraftsGenerated)
}

// HeuristicDFlashDraft wraps a target model for lightweight draft heuristics
// when a full DFlashDraftModel is not loaded.
type HeuristicDFlashDraft struct {
	Target Model
	Config DFlashConfig
	Decode DFlashDecodeOpts
	Stats  DFlashStats
	Cache  []Token
}

// NewHeuristicDFlashDraft constructs a heuristic draft around a target model.
func NewHeuristicDFlashDraft(target Model, config DFlashConfig) *HeuristicDFlashDraft {
	return NewHeuristicDFlashDraftWithOpts(target, config, DefaultDFlashDecodeOpts())
}

// NewHeuristicDFlashDraftWithOpts constructs a heuristic draft with decode tuning.
func NewHeuristicDFlashDraftWithOpts(target Model, config DFlashConfig, decode DFlashDecodeOpts) *HeuristicDFlashDraft {
	if decode.DraftSteps <= 0 {
		if config.BlockSize > 0 {
			decode.DraftSteps = config.BlockSize
		} else {
			decode.DraftSteps = 4
		}
	}
	return &HeuristicDFlashDraft{Target: target, Config: config, Decode: decode}
}

// Forward implements Model by delegating to the target.
func (d *HeuristicDFlashDraft) Forward(tokens []Token, session *Session) (Logits, error) {
	return d.Target.Forward(tokens, session)
}

// VocabSize delegates to the target model.
func (d *HeuristicDFlashDraft) VocabSize() int { return d.Target.VocabSize() }

// ContextSize delegates to the target model.
func (d *HeuristicDFlashDraft) ContextSize() int { return d.Target.ContextSize() }

// LayerCount delegates to the target model.
func (d *HeuristicDFlashDraft) LayerCount() int { return d.Target.LayerCount() }

// Reset clears the draft cache.
func (d *HeuristicDFlashDraft) Reset() { d.Cache = d.Cache[:0] }

// GenerateDraft produces a draft of size steps tokens.
func (d *HeuristicDFlashDraft) GenerateDraft(prompt []Token, steps int) ([]Token, error) {
	if steps <= 0 {
		steps = d.Decode.DraftSteps
	}
	out := append([]Token(nil), prompt...)
	sess := NewSession()
	for i := 0; i < steps; i++ {
		logits, err := d.Target.Forward(out, sess)
		if err != nil {
			return nil, err
		}
		tok, err := Sample(logits, SamplingConfig{
			Temperature: d.Decode.Temperature,
			TopP:        1.0,
		}, nil)
		if err != nil {
			return nil, err
		}
		out = append(out, tok)
		d.Cache = append(d.Cache, tok)
	}
	d.Stats.DraftsGenerated += steps
	return out[len(prompt):], nil
}

// VerifyDrafts runs a verification pass and accepts drafts above MinDraftConfidence.
func (d *HeuristicDFlashDraft) VerifyDrafts(drafts []Token) (accepted []Token, err error) {
	if len(drafts) == 0 {
		return nil, nil
	}
	sess := NewSession()
	logits, err := d.Target.Forward(drafts, sess)
	if err != nil {
		return nil, err
	}
	probs := softmax(logits)
	for _, t := range drafts {
		prob := float32(0)
		if int(t) < len(probs) {
			prob = probs[t]
		}
		if prob >= d.Decode.MinDraftConfidence {
			accepted = append(accepted, t)
		}
	}
	d.Stats.DraftsAccepted += len(accepted)
	return accepted, nil
}

// DraftConfidence computes per-token confidence for the draft sequence.
func DraftConfidence(target Model, drafts []Token) ([]float32, error) {
	if len(drafts) == 0 {
		return nil, nil
	}
	sess := NewSession()
	logits, err := target.Forward(drafts, sess)
	if err != nil {
		return nil, err
	}
	probs := softmax(logits)
	out := make([]float32, len(drafts))
	for i, t := range drafts {
		if int(t) < len(probs) {
			out[i] = probs[t]
		}
	}
	return out, nil
}

// DFlashEngine orchestrates heuristic draft + target speculative steps.
type DFlashEngine struct {
	Target Model
	Draft  *HeuristicDFlashDraft
	Config DFlashConfig
	Decode DFlashDecodeOpts
	Stats  DFlashStats
}

// NewDFlashEngine constructs a draft + target engine.
func NewDFlashEngine(target Model, draft *HeuristicDFlashDraft, config DFlashConfig) *DFlashEngine {
	decode := DefaultDFlashDecodeOpts()
	if draft != nil {
		decode = draft.Decode
	}
	return &DFlashEngine{Target: target, Draft: draft, Config: config, Decode: decode}
}

// Step runs a single DFlash decode iteration.
func (e *DFlashEngine) Step(prompt []Token) (accepted []Token, err error) {
	drafts, err := e.Draft.GenerateDraft(prompt, e.Decode.DraftSteps)
	if err != nil {
		return nil, err
	}
	accepted, err = e.Draft.VerifyDrafts(drafts)
	if err != nil {
		return nil, err
	}
	ratio := e.Stats.AcceptanceRate()
	e.Stats.SpeedupEstimate = float32(1 + math.Ceil(float64(e.Decode.DraftSteps)*float64(ratio)))
	return accepted, nil
}

// MaxAcceptanceRate returns the configured maximum acceptance rate.
func (e *DFlashEngine) MaxAcceptanceRate() float32 { return e.Decode.MaxAcceptanceRate }

// IsStable returns true when acceptance rate is within configured bounds.
func (e *DFlashEngine) IsStable() bool {
	return e.Stats.AcceptanceRate() <= e.Decode.MaxAcceptanceRate
}
